### PR TITLE
Feature/Web/Protected Components

### LINKED ISSUE #119: Criar Rotas Protegidas
## Feature Description

Criar um componente "Protected" para que possamos proteger rotas e/ou componentes. 
Exemplo de uso:

**Proteger por estado de autenticação (só renderiza os filhos caso esteja logado).**

```javascript
<Protected>
	<FormWizard
		onSubmit={handleSubmit}
		onPrev={({ prevStep }) => setCurrentStep(prevStep)}
		currentStep={currentStep}
		steps={newTechonologySteps}
	/>
</Protected>
```

**Proteger por `role`.**
Renderize `children` caso o usuário esteja logado e possua a `role` passada.

```javascript
<Protected role="admin">
	<Components />
</Protected>
```

Por padrão caso o usuário não esteja logado, será mostrado o modal de login, no lugar de `children`. Após fazer o login caso o usuário tenha permissão ele poderá então visualizar `children`.

Caso o usuário não tenha permissão (e esteja logado) deverá ser renderizado um componente `<NotAuthorized />` com uma mensagem padrão (design a ser definido).

O componente `Protected`  também poderá receber (de forma opcional) um `redirectTo` para que caso o usuário não tenha permissão (após logado) ele ser redirecionado para a rota específicada em `redtirectTo`.

---------------

## Acceptance criteria

* `<Protected>` não renderiza `children` caso o usuário não esteja logado ou não tenha permissão.
* `<Protected>` não renderiza `children` e mostra o formulário de login caso o usuário não esteja logado.
* `<Protected>` não renderiza `children` e mostra o formulário de login caso o usuário não esteja logado. E após se logar e o usuário tiver permissão renderiza `children`
* `<Protected role="admin" redirectTo="/">` não renderiza `children` e redireciona o usuário para `redirectTo` caso ele **esteja logado** e não possua permissão.
* `<Protected role="admin" ` não renderiza `children` e exibte o componente `NotAuthorized` ocaso ele **esteja logado** e não possua permissão.

## Implementation Brief

* Criar o componente `NotAuthorized.js` em `packages/web/components/Authorization/`
* Criar o componente `Protected.js` em  `packages/web/components/Authorization/`
* Exportar ambos os componentes em  `packages/web/components/Authorization/index.js`
* `Protected.js` deverá usar o hook `useUser` para verificar se o usuário está logado e suas devidas permissões.

Exemplo:
```javascript
const Protected = ({ children, redirectTo, role }) => {
   const user = useUser();
   
   if (!user) {
     // show login modal.
   }

   if (role && user.role !== role) {
      if (redirectTo) {
        // redireciona
      } else {
         return <NotAuthorized />
      }
   }

   return (<>{children}</>)
}
```

### PR DESCRIPTION
**Pull request by mateus4k**
## Summary

A component to protect another components from unauthorized access.

Resolves #119.

## Relevant technical choices

- By default, the component protects children only from not logged in users:
```jsx
const App = () => {
	<Protected>
		<h1>Protected Children</h1>
	</Protected>
};
```

- We can also protect children based on authorization using the `role` prop and pass a URL to redirect to another page:
```jsx
const App = () => {
	<Protected role="admin" redirectTo="/about">
		<h1>Protected Children</h1>
	</Protected>
};
```

## QA Steps

Only check if user is logged in:

basic usage

Pass a needed user role:

user role

Redirect to another page:

redirect

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [ ] My code has proper inline documentation.
- [x] I have manually tested this PR.

### CODE REVIEW

File: packages/web/components/Authorization/Protected.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+		if (role && user.role !== role) {
+			if (redirectTo) {
+				return Router.push(redirectTo);
+			}
+			return <NotAuthorized />;
+		}
+
+		return <>{children}</>;
+	});
```
Comment (nicholasio) [inline packages/web/components/Authorization/Protected.js:28]: Você precisa passar as depêndencias do `useEffect`, caso contrário ele faz executar em todo re-render. Certifique-se que o eslint tá rodando junto com o vscode pois ele te mostra essas coisas!

Aqui você precisa de email, role e user, além do openModal.
```javascript
useEffect(() => {
		if (!user?.email) {
			return openModal('login', {
				message: t('common:signInToContinue'),
			});
		}

		if (role && user.role !== role) {
			if (redirectTo) {
				return Router.push(redirectTo);
			}
			return <NotAuthorized />;
		}

		return <>{children}</>;
	}, [user, openModal, role]);
```

File: packages/web/components/Authorization/Protected.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	useEffect(() => {
+		if (!user?.email) {
+			return openModal('login', {
+				message: t('common:signInToContinue'),
+			});
+		}
+
+		if (role && user.role !== role) {
+			if (redirectTo) {
+				return Router.push(redirectTo);
```
Comment (nicholasio) [inline packages/web/components/Authorization/Protected.js:22]: Use o `useRouter` https://nextjs.org/docs/api-reference/next/router#userouter

`const router = useRouter()`;

e depois

`router.push(redirectTo);`

File: packages/web/components/Authorization/Protected.js (hunk 1 above)
Comment (mateus4k) [reply packages/web/components/Authorization/Protected.js:28]: Bem observado! Não pensei em passar as dependências pois não imaginei que seriam possíveis ou necessários rerenders, mas agora faz sentido pra mim pensar nisso. Nesse caso as dependências funcionariam como um shouldComponentUpdate, correto? Vou fazer a correção!

File: packages/web/components/Authorization/Protected.js (hunk 2 above)
Comment (mateus4k) [reply packages/web/components/Authorization/Protected.js:22]: Valeu! Ainda não tinha testado essa function! 🙂

**Comment by mateus4k** [comment]
Reviewers, you can disregard the `UserProvider.js` conflict. This change was just a hotfix.

**Pull request merged**